pkg/controller/certificaterequests/acme/acme.go

```diff
@@ -165,37 +165,47 @@ func (a *ACME) Sign(ctx context.Context, cr *v1.CertificateRequest, issuer v1.Ge
 		return nil, nil
 	}
 
-	// Order valid, return cert. The calling controller will update with ready if its happy with the cert.
-	if order.Status.State == cmacme.Valid {
-		x509Cert, err := pki.DecodeX509CertificateBytes(order.Status.Certificate)
-		if errors.IsInvalidData(err) {
-			log.Error(err, "failed to decode x509 certificate data on Order resource")
-			return nil, a.acmeClientV.Orders(order.Namespace).Delete(context.TODO(), order.Name, metav1.DeleteOptions{})
-		}
-		ok, err := pki.PublicKeyMatchesCertificate(csr.PublicKey, x509Cert)
-		if err != nil {
-			return nil, err
-		}
-		if !ok {
-			log.Error(err, "failed to decode x509 certificate data on Order resource, recreating...")
-			return nil, a.acmeClientV.Orders(order.Namespace).Delete(context.TODO(), order.Name, metav1.DeleteOptions{})
-		}
+	if order.Status.State != cmacme.Valid {
+		// We update here to just pending while we wait for the order to be resolved.
+		a.reporter.Pending(cr, nil, "OrderPending",
+			fmt.Sprintf("Waiting on certificate issuance from order %s/%s: %q",
+				expectedOrder.Namespace, order.Name, order.Status.State))
 
-		log.V(logf.InfoLevel).Info("certificate issued")
+		log.V(logf.DebugLevel).Info("acme Order resource is not in a ready state, waiting...")
 
-		return &issuerpkg.IssueResponse{
-			Certificate: order.Status.Certificate,
-		}, nil
+		return nil, nil
 	}
 
-	// We update here to just pending while we wait for the order to be resolved.
-	a.reporter.Pending(cr, nil, "OrderPending",
-		fmt.Sprintf("Waiting on certificate issuance from order %s/%s: %q",
-			expectedOrder.Namespace, order.Name, order.Status.State))
+	if len(order.Status.Certificate) == 0 {
+		a.reporter.Pending(cr, nil, "OrderPending",
+			fmt.Sprintf("Waiting for order-controller to add certificate data to Order %s/%s",
+				expectedOrder.Namespace, order.Name))
+
+		log.V(logf.DebugLevel).Info("Order controller has not added certificate data to the Order, waiting...")
+		return nil, nil
+	}
 
-	log.V(logf.DebugLevel).Info("acme Order resource is not in a ready state, waiting...")
+	// Order valid, return cert. The calling controller will update with ready if its happy with the cert.
+	x509Cert, err := pki.DecodeX509CertificateBytes(order.Status.Certificate)
+	if errors.IsInvalidData(err) {
+		log.Error(err, "failed to decode x509 certificate data on Order resource")
+		return nil, a.acmeClientV.Orders(order.Namespace).Delete(context.TODO(), order.Name, metav1.DeleteOptions{})
+	}
+	ok, err := pki.PublicKeyMatchesCertificate(csr.PublicKey, x509Cert)
+	if err != nil {
+		return nil, err
```

Comment:
We handle all error cases with a retry of the order, why don't we do it on this error?

Reply:
I don't know. The error was always returned in this case.
But I agree that it seems fishy.
If the Certificate doesn't match the public key, then that strikes me as a permanent error and there's no point retrying this CertificateRequest. 
It should be marked as failed.

Reply:
I see but i could say the same about invalid x509 data

Reply:
Good point @meyskens. I'd misunderstood your comment and I hadn't investigated the `PublicKeyMatchesCertificate` code.
It does a type switch on the supplied `PublicKey` and returns an error if the key has an unhandled type.

I've now updated the controller to delete the order in both cases: when there is a mismatch *and* if there is an unhandled key type.

Also added tests for the first of these and for the case where the Order certificate is not valid x509 PEM bytes